In [1]:
import pandas as pd
import numpy as np

In [2]:
def standard_headers(df):
    cols = [col.lower().replace(' ', '_') for col in df.columns]
    df.columns = cols
    return df

In [3]:
airplanes = pd.read_csv('../data/raw/Airplane_Crashes_and_Fatalities_Since_1908.csv')
airplanes = standard_headers(airplanes)
airplanes = airplanes.rename(columns={'flight_#': 'flight_no'}) #Character can be a problem so we get rid of it.
display(airplanes.head())
display(airplanes.shape)

date   time                            location  \
0  09/17/1908  17:18                 Fort Myer, Virginia   
1  07/12/1912  06:30             AtlantiCity, New Jersey   
2  08/06/1913    NaN  Victoria, British Columbia, Canada   
3  09/09/1913  18:30                  Over the North Sea   
4  10/17/1913  10:30          Near Johannisthal, Germany   

                 operator flight_no          route                    type  \
0    Military - U.S. Army       NaN  Demonstration        Wright Flyer III   
1    Military - U.S. Navy       NaN    Test flight               Dirigible   
2                 Private         -            NaN        Curtiss seaplane   
3  Military - German Navy       NaN            NaN  Zeppelin L-1 (airship)   
4  Military - German Navy       NaN            NaN  Zeppelin L-2 (airship)   

  registration cn/in  aboard  fatalities  ground  \
0          NaN     1     2.0         1.0     0.0   
1          NaN   NaN     5.0         5.0     0.0   
2          NaN   NaN     1.0         1.0     0.0   
3          NaN   NaN    20.0        14.0     0.0   
4          NaN   NaN    30.0        30.0     0.0   

                                             summary  
0  During a demonstration flight, a U.S. Army fly...  
1  First U.S. dirigible Akron exploded just offsh...  
2  The first fatal airplane accident in Canada oc...  
3  The airship flew into a thunderstorm and encou...  
4  Hydrogen gas which was being vented was sucked...

(5268, 13)

In [4]:
airplanes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5268 entries, 0 to 5267
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   date          5268 non-null   object 
 1   time          3049 non-null   object 
 2   location      5248 non-null   object 
 3   operator      5250 non-null   object 
 4   flight_no     1069 non-null   object 
 5   route         3562 non-null   object 
 6   type          5241 non-null   object 
 7   registration  4933 non-null   object 
 8   cn/in         4040 non-null   object 
 9   aboard        5246 non-null   float64
 10  fatalities    5256 non-null   float64
 11  ground        5246 non-null   float64
 12  summary       4878 non-null   object 
dtypes: float64(3), object(10)
memory usage: 535.2+ KB


In [5]:
airplanes.isna().sum()

date               0
time            2219
location          20
operator          18
flight_no       4199
route           1706
type              27
registration     335
cn/in           1228
aboard            22
fatalities        12
ground            22
summary          390
dtype: int64

We can observe quite a number of Nas. This could be due to multiple reasons: lack of obligation to provide that information (as many measures were implemented with time and progressively with the industry development), non-registered flights (like military or pirate flights), unclear information or damaged sensors.

For now we will not drop any rows or columns and continue to explore the data, as one missing value in a row may not be that relevant to understand the case. Date is the essential paramenter to classify the flights (also will allow us to build a time accident rate) and we have no missing values on that one.

<h3>Let's explore the location column.<h3>

In [6]:
airplanes['location'].nunique(dropna = False)

4304

In [7]:
airplanes['location'].value_counts()

Sao Paulo, Brazil                    15
Moscow, Russia                       15
Rio de Janeiro, Brazil               14
Anchorage, Alaska                    13
Manila, Philippines                  13
                                     ..
Near Charana, Bolivia                 1
Monte Matto, Italy                    1
Misaki Mountain, Japan                1
Angelholm, Sweden                     1
State of Arunachal Pradesh, India     1
Name: location, Length: 4303, dtype: int64

We have 4304 different locations for the flights (we also have 20 Nas and one of them is included in this calculation, so in reality we would have 4303).

Some of these locations can be in the same country and even region or city. As this column stores info as a string, just a minor difference in the writing would count as a different value, so we will have to consider that when using the information (as we can see in the value counts - 2 locations from the same country on top, an USA state but without stating the country). We will have to get the country information from this column.

<h3>Let's now check the flight operators:<h3>

In [8]:
airplanes['operator'].nunique()

2476

In [9]:
airplanes['operator'].value_counts()

Aeroflot                               179
Military - U.S. Air Force              176
Air France                              70
Deutsche Lufthansa                      65
Air Taxi                                44
                                      ... 
Military - Argentine Navy                1
Richland Flying Service - Air Taxii      1
Harbor Airlines - Air Taxi               1
Aerovias Venezolanas SA (Venezuela)      1
Strait Air                               1
Name: operator, Length: 2476, dtype: int64

We have 2476 unique flight operators. Nas have not been considered as the nunique function drops them by default. Here we have some considerations: this info is also stored in a string so any minor difference or typo would count as a different value. Also, some flights are labeled as private - they may be grouped but it doesn't mean they were handled by the same company.
We can already see which operators have the biggest incident rate: Aeroflot, USAF and Air France (closely followed by Lufthansa).

<h3>Let's check the flight number column:<h3>

In [10]:
airplanes['flight_no'].value_counts(dropna = False)

NaN     4199
-         67
1         10
4          7
21         6
        ... 
621        1
215        1
208B       1
158        1
447        1
Name: flight_no, Length: 725, dtype: int64

There are many Nas, it may be interesting to keep it just for extra information purposes (for example, to provide extra narrative context when talking about an specific flight when this info is available). That would be the only purpose of this column, also because flight names are usually based on aircraft routes, and these can change with relative frequence depending on many factors.

<h3>Let's check now the route column:<h3>

In [11]:
airplanes['route'].value_counts(dropna = False)

NaN                           1706
Training                        81
Sightseeing                     29
Test flight                     17
Test                             6
                              ... 
Manila - Lapu Lapu               1
Saint Denis - Paris              1
Cork - London                    1
Peoria, IL - St. Louis, MO       1
Mechuka for Jorhat               1
Name: route, Length: 3245, dtype: int64

As before, we have quite a number of Nas but it may be interesting to keep this column for extra context.

<h3>We will check now the type column, which indicates the plane model that was involved in the accident:<h3>

In [12]:
airplanes['type'].value_counts(dropna = False)

Douglas DC-3                                334
de Havilland Canada DHC-6 Twin Otter 300     81
Douglas C-47A                                74
Douglas C-47                                 62
Douglas DC-4                                 40
                                           ... 
Boeing 727-21                                 1
NAMC-YS-11-111                                1
Lockheed EC-121H                              1
Cessna 205A                                   1
Airbus A330-203                               1
Name: type, Length: 2447, dtype: int64

We can observe that the Douglas models have the highest number of accidents by far. We can later consider to cross check it with the operators.

<h3>registration and cn/In<h3>

These columns indicate the registration number of the aircraft and the serial number given by the manufacturer. It is not relevant for our dataset and they contain quite a number or Nas, but we will keep them as we did with the flight number.

<h3>Aboard<h3>

Time to check the 'aboard' column, which indicates how many persons were aboard each flight (including crew members as the lower values indicate).

In [13]:
airplanes['aboard'].value_counts(dropna = False)

2.0      377
3.0      370
4.0      296
5.0      239
6.0      223
        ... 
240.0      1
218.0      1
192.0      1
269.0      1
228.0      1
Name: aboard, Length: 240, dtype: int64

In [14]:
airplanes['aboard'].isna().sum()

22

There are 22 flights for which we have no information on how many passengers were onboard. We can assume that there was at least the pilot on board, so we will check the type of flight to see if we can gather more information and eventually fill in the values by comparing them to similar flights.<br> As a last resort, we could also look for them on google as the information could be available, and we have enough data to do a search.<br>
We will also take advantage of the descritpion column to get more information.

In [15]:
airplanes_nopassengers = airplanes[(airplanes['aboard'].isna())]
airplanes_nopassengers

date time                                 location  \
26    10/20/1919  NaN                          English Channel   
333   08/10/1934  NaN                            Ningbo, China   
348   03/07/1935  NaN                    Schievelbein, Germany   
364   08/13/1935  NaN                            Hangow, China   
423   12/26/1936  NaN                           Nanking, China   
526   09/26/1939  NaN                                North Sea   
537   07/07/1940  NaN                           Gulf of Tonkin   
570   01/24/1942  NaN                   Near Samarinda, Borneo   
571   01/26/1942  NaN                            Kupang, Timor   
573   02/14/1942  NaN                                      NaN   
587   08/13/1942  NaN                         Algiers, Algeria   
593   10/01/1942  NaN                           Kunming, China   
678   11/09/1944  NaN                          Seljord, Norway   
768   03/18/1946  NaN    Between Chungking and Shanghai, China   
832   12/25/1946  NaN                 Lunghwa, Shanghai, China   
1479  04/20/1957  NaN                            Jirkouk, Iraq   
3007  11/03/1977  NaN                     Belgrade, Yugoslavia   
3323  12/16/1981  NaN                     Kuala Belait, Brunei   
3369  08/11/1982  NaN                              PacifiOcean   
3843  05/09/1989  NaN                      Near Tainjin, China   
4080  02/20/1992  NaN  En route from Argentina  to  California   
4705  03/22/2000  NaN                          Herreira, Spain   

                                 operator flight_no  \
26          Aircraft Transport and Travel       NaN   
333   China National Aviation Corporation       NaN   
348                              Deruluft       NaN   
364   China National Aviation Corporation       NaN   
423   China National Aviation Corporation       NaN   
526              KLM Royal Dutch Airlines       NaN   
537                            Air France       NaN   
570                                 KNILM       NaN   
571                                 KNILM       NaN   
573   China National Aviation Corporation       NaN   
587                            Air France       NaN   
593   China National Aviation Corporation       NaN   
678        Military - U.S. Army Air Corps       NaN   
768   China National Aviation Corporation       NaN   
832   China National Aviation Corporation       NaN   
1479                           Air France       NaN   
3007                                El Al       NaN   
3323                  Bristow Helicopters       NaN   
3369           Pan American World Airways       830   
3843                           Aero Asahi       NaN   
4080                Aerolineas Argentinas       386   
4705         Military - Ejército del Aire       NaN   

                                route                             type  \
26                                NaN                De Havilland DH-4   
333                               NaN                   Sikorsky S-38B   
348                               NaN                 Rochrbach Roland   
364                               NaN                   Sikorsky S-38B   
423                               NaN                     Douglas DC-2   
526             Stockholm - Amsterdam                     Douglas DC-3   
537                               NaN                  Dewoitine D-338   
570                               NaN                     Douglas DC-3   
571                               NaN               Grumman G-21 Goose   
573                               NaN                     Douglas DC-2   
587                               NaN  Liore et Olivier H-246 Air Boat   
593                               NaN                     Douglas C-47   
678                               NaN                              NaN   
768               Chunking - Shanghai                              NaN   
832                               NaN         Curtiss C-46, C-47, DC-3   
1479  Tehran, Iran - Istanbul, Turkey     L

In [16]:
airplanes['summary'][526]

'One Swedish passenger was killed when the plane was attacked by German fighters. The plane was able to land safely in Amsterdam.'

In [17]:
airplanes['summary'][570]

'Shot down by Japanese military aircraft.'

In [18]:
airplanes['summary'][1479]

'Explosive decompression. A passenger was sucked out of a window at 18,000 feet when the window he was sitting next to shattered. The body was never recovered.'

In [19]:
airplanes['summary'][3369]

'A sixteen-year-old boy was killed when a bomb detonated under a seat cushion. The explosion caused minor damage and the plane landed safely at Honolulu. The bomb was placed onboard by Mohammed Rashed, a Jordanian terrorist with the May 15 Organization.'

-8 aircrafts were involved on incidents during the WWII period. Some of them directly involved in combat (three of them shot down by a Japanese military fighter). Other on which a passenger died after the plane was attacked by a fighter.
-One of the incidents refers to a passenger dying of cholera while en route.
-One of the incidents involves a bomb exploding on board and killing a passenger, but the plane managed to land without any further casualties.

Since we want to focus exclusively on aircraft accidents - whether these are mechanic, man-cause or weather caused - we can drop these rows as non relevant. This also could mean that some of the casualties in the dataset are not related to airplane crashes, but on-flight health related deaths. We will have to dig in deeper but we will start by clearing these rows (7 in total).

In [20]:
airplanes = airplanes.drop([airplanes.index[526], airplanes.index[537], airplanes.index[570], airplanes.index[571], airplanes.index[587], airplanes.index[3369], airplanes.index[4080]])

In [21]:
airplanes = airplanes.reset_index(drop=True)

In [22]:
airplanes_nopassengers = airplanes[(airplanes['aboard'].isna())]
airplanes_nopassengers

date time                               location  \
26    10/20/1919  NaN                        English Channel   
333   08/10/1934  NaN                          Ningbo, China   
348   03/07/1935  NaN                  Schievelbein, Germany   
364   08/13/1935  NaN                          Hangow, China   
423   12/26/1936  NaN                         Nanking, China   
569   02/14/1942  NaN                                    NaN   
588   10/01/1942  NaN                         Kunming, China   
673   11/09/1944  NaN                        Seljord, Norway   
763   03/18/1946  NaN  Between Chungking and Shanghai, China   
827   12/25/1946  NaN               Lunghwa, Shanghai, China   
1474  04/20/1957  NaN                          Jirkouk, Iraq   
3002  11/03/1977  NaN                   Belgrade, Yugoslavia   
3318  12/16/1981  NaN                   Kuala Belait, Brunei   
3837  05/09/1989  NaN                    Near Tainjin, China   
4698  03/22/2000  NaN                        Herreira, Spain   

                                 operator flight_no  \
26          Aircraft Transport and Travel       NaN   
333   China National Aviation Corporation       NaN   
348                              Deruluft       NaN   
364   China National Aviation Corporation       NaN   
423   China National Aviation Corporation       NaN   
569   China National Aviation Corporation       NaN   
588   China National Aviation Corporation       NaN   
673        Military - U.S. Army Air Corps       NaN   
763   China National Aviation Corporation       NaN   
827   China National Aviation Corporation       NaN   
1474                           Air France       NaN   
3002                                El Al       NaN   
3318                  Bristow Helicopters       NaN   
3837                           Aero Asahi       NaN   
4698         Military - Ejército del Aire       NaN   

                                route                          type  \
26                                NaN             De Havilland DH-4   
333                               NaN                Sikorsky S-38B   
348                               NaN              Rochrbach Roland   
364                               NaN                Sikorsky S-38B   
423                               NaN                  Douglas DC-2   
569                               NaN                  Douglas DC-2   
588                               NaN                  Douglas C-47   
673                               NaN                           NaN   
763               Chunking - Shanghai                           NaN   
827                               NaN      Curtiss C-46, C-47, DC-3   
1474  Tehran, Iran - Istanbul, Turkey  Lockheed Super Constellation   
3002                              NaN                  Boeing B-747   
3318                              NaN             Aerospatiale Puma   
3837                              NaN                      Bell 412   
4698               Sevilla - Herreira       CASA 212-DE Aviocar 200   

     registration  cn/in  aboard  fatalities  ground  \
26         G-EAHG    NaN     NaN         NaN     NaN   
333           NaN    NaN     NaN         NaN     NaN   
348        D-AJYP     45     NaN         3.0     0.0   
364         NV40V    NaN     NaN         NaN     NaN   
423       NC14269    NaN     NaN         NaN     NaN   
569            45    NaN     NaN         NaN     NaN   
588            69    NaN     NaN         NaN     NaN   
673      42-52196    NaN     NaN         NaN     NaN   
763           139    NaN     NaN         NaN     NaN   
827           115    NaN     NaN        87.0     4.0   
1474       F-BGNE   4514     NaN         1.0     0.0   
3002          NaN    NaN     NaN         1.0     0.0   
3318       9M-SSC   1481     NaN        12.0     0.0   
3837       JA9596  33070     NaN        10.0     0.0   
4698    TM-12D-73    314     NaN         NaN     NaN   

                                                summary  
26            

The last flight of the list is quite recent so we will probably be able to find more information on google.

7 people dead.
https://hemeroteca.lavanguardia.com/preview/2000/03/23/pagina-40/34056506/pdf.html?search=aviocar

In [23]:
airplanes.iloc[4698]

date                                                   03/22/2000
time                                                          NaN
location                                          Herreira, Spain
operator                             Military - Ejército del Aire
flight_no                                                     NaN
route                                          Sevilla - Herreira
type                                      CASA 212-DE Aviocar 200
registration                                            TM-12D-73
cn/in                                                         314
aboard                                                        NaN
fatalities                                                    NaN
ground                                                        NaN
summary         Crashed while attempting to land in poor weather.
Name: 4698, dtype: object

In [24]:
airplanes['aboard'][4698] = 7

/tmp/ipykernel_45867/1304270209.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['aboard'][4698] = 7


In [25]:
airplanes['fatalities'][4698] = 7

/tmp/ipykernel_45867/826404550.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['fatalities'][4698] = 7


In [26]:
airplanes.iloc[4698]

date                                                   03/22/2000
time                                                          NaN
location                                          Herreira, Spain
operator                             Military - Ejército del Aire
flight_no                                                     NaN
route                                          Sevilla - Herreira
type                                      CASA 212-DE Aviocar 200
registration                                            TM-12D-73
cn/in                                                         314
aboard                                                        7.0
fatalities                                                    7.0
ground                                                        NaN
summary         Crashed while attempting to land in poor weather.
Name: 4698, dtype: object

In [27]:
airplanes_nopassengers = airplanes[(airplanes['aboard'].isna())]
airplanes_nopassengers

date time                               location  \
26    10/20/1919  NaN                        English Channel   
333   08/10/1934  NaN                          Ningbo, China   
348   03/07/1935  NaN                  Schievelbein, Germany   
364   08/13/1935  NaN                          Hangow, China   
423   12/26/1936  NaN                         Nanking, China   
569   02/14/1942  NaN                                    NaN   
588   10/01/1942  NaN                         Kunming, China   
673   11/09/1944  NaN                        Seljord, Norway   
763   03/18/1946  NaN  Between Chungking and Shanghai, China   
827   12/25/1946  NaN               Lunghwa, Shanghai, China   
1474  04/20/1957  NaN                          Jirkouk, Iraq   
3002  11/03/1977  NaN                   Belgrade, Yugoslavia   
3318  12/16/1981  NaN                   Kuala Belait, Brunei   
3837  05/09/1989  NaN                    Near Tainjin, China   

                                 operator flight_no  \
26          Aircraft Transport and Travel       NaN   
333   China National Aviation Corporation       NaN   
348                              Deruluft       NaN   
364   China National Aviation Corporation       NaN   
423   China National Aviation Corporation       NaN   
569   China National Aviation Corporation       NaN   
588   China National Aviation Corporation       NaN   
673        Military - U.S. Army Air Corps       NaN   
763   China National Aviation Corporation       NaN   
827   China National Aviation Corporation       NaN   
1474                           Air France       NaN   
3002                                El Al       NaN   
3318                  Bristow Helicopters       NaN   
3837                           Aero Asahi       NaN   

                                route                          type  \
26                                NaN             De Havilland DH-4   
333                               NaN                Sikorsky S-38B   
348                               NaN              Rochrbach Roland   
364                               NaN                Sikorsky S-38B   
423                               NaN                  Douglas DC-2   
569                               NaN                  Douglas DC-2   
588                               NaN                  Douglas C-47   
673                               NaN                           NaN   
763               Chunking - Shanghai                           NaN   
827                               NaN      Curtiss C-46, C-47, DC-3   
1474  Tehran, Iran - Istanbul, Turkey  Lockheed Super Constellation   
3002                              NaN                  Boeing B-747   
3318                              NaN             Aerospatiale Puma   
3837                              NaN                      Bell 412   

     registration  cn/in  aboard  fatalities  ground  \
26         G-EAHG    NaN     NaN         NaN     NaN   
333           NaN    NaN     NaN         NaN     NaN   
348        D-AJYP     45     NaN         3.0     0.0   
364         NV40V    NaN     NaN         NaN     NaN   
423       NC14269    NaN     NaN         NaN     NaN   
569            45    NaN     NaN         NaN     NaN   
588            69    NaN     NaN         NaN     NaN   
673      42-52196    NaN     NaN         NaN     NaN   
763           139    NaN     NaN         NaN     NaN   
827           115    NaN     NaN        87.0     4.0   
1474       F-BGNE   4514     NaN         1.0     0.0   
3002          NaN    NaN     NaN         1.0     0.0   
3318       9M-SSC   1481     NaN        12.0     0.0   
3837       JA9596  33070     NaN        10.0     0.0   

                                                summary  
26                                                  NaN  
333                                                 NaN  
348                                   Fuselage failure.  
364                               Destoryed in a storm.  
423                         

For the last value (05/09/1989 Near Tainjin, China), 10/10 dead:
https://aviation-safety.net/wikibase/32772

In [28]:
airplanes['aboard'][3837] = 10

/tmp/ipykernel_45867/3195782740.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['aboard'][3837] = 10


In [29]:
airplanes['fatalities'][3837] = 10

/tmp/ipykernel_45867/1402348532.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['fatalities'][3837] = 10


In [30]:
airplanes_nopassengers = airplanes[(airplanes['aboard'].isna())]
airplanes_nopassengers

date time                               location  \
26    10/20/1919  NaN                        English Channel   
333   08/10/1934  NaN                          Ningbo, China   
348   03/07/1935  NaN                  Schievelbein, Germany   
364   08/13/1935  NaN                          Hangow, China   
423   12/26/1936  NaN                         Nanking, China   
569   02/14/1942  NaN                                    NaN   
588   10/01/1942  NaN                         Kunming, China   
673   11/09/1944  NaN                        Seljord, Norway   
763   03/18/1946  NaN  Between Chungking and Shanghai, China   
827   12/25/1946  NaN               Lunghwa, Shanghai, China   
1474  04/20/1957  NaN                          Jirkouk, Iraq   
3002  11/03/1977  NaN                   Belgrade, Yugoslavia   
3318  12/16/1981  NaN                   Kuala Belait, Brunei   

                                 operator flight_no  \
26          Aircraft Transport and Travel       NaN   
333   China National Aviation Corporation       NaN   
348                              Deruluft       NaN   
364   China National Aviation Corporation       NaN   
423   China National Aviation Corporation       NaN   
569   China National Aviation Corporation       NaN   
588   China National Aviation Corporation       NaN   
673        Military - U.S. Army Air Corps       NaN   
763   China National Aviation Corporation       NaN   
827   China National Aviation Corporation       NaN   
1474                           Air France       NaN   
3002                                El Al       NaN   
3318                  Bristow Helicopters       NaN   

                                route                          type  \
26                                NaN             De Havilland DH-4   
333                               NaN                Sikorsky S-38B   
348                               NaN              Rochrbach Roland   
364                               NaN                Sikorsky S-38B   
423                               NaN                  Douglas DC-2   
569                               NaN                  Douglas DC-2   
588                               NaN                  Douglas C-47   
673                               NaN                           NaN   
763               Chunking - Shanghai                           NaN   
827                               NaN      Curtiss C-46, C-47, DC-3   
1474  Tehran, Iran - Istanbul, Turkey  Lockheed Super Constellation   
3002                              NaN                  Boeing B-747   
3318                              NaN             Aerospatiale Puma   

     registration cn/in  aboard  fatalities  ground  \
26         G-EAHG   NaN     NaN         NaN     NaN   
333           NaN   NaN     NaN         NaN     NaN   
348        D-AJYP    45     NaN         3.0     0.0   
364         NV40V   NaN     NaN         NaN     NaN   
423       NC14269   NaN     NaN         NaN     NaN   
569            45   NaN     NaN         NaN     NaN   
588            69   NaN     NaN         NaN     NaN   
673      42-52196   NaN     NaN         NaN     NaN   
763           139   NaN     NaN         NaN     NaN   
827           115   NaN     NaN        87.0     4.0   
1474       F-BGNE  4514     NaN         1.0     0.0   
3002          NaN   NaN     NaN         1.0     0.0   
3318       9M-SSC  1481     NaN        12.0     0.0   

                                                summary  
26                                                  NaN  
333                                                 NaN  
348                                   Fuselage failure.  
364                               Destoryed in a storm.  
423                                                 NaN  
569                                                 NaN  
588   Crashed while attempting to land after losing ...  
673                                                 NaN  
763    Disappeared while en route. Plane never located. 

10/20/1919 - http://www.planecrashinfo.com/1920/1920-31.htm One aboard, no fatalities.<br>
08/10/1934 - http://www.planecrashinfo.com/1934/1934-18.htm Undetermined aboard & fatalities.<br>
03/07/1935 - https://aviation-safety.net/database/record.php?id=19350720-0 13/13 dead<br>
08/13/1935 - http://www.planecrashinfo.com/1935/1935-24.htm Undetermined aboard & fatalities.<br>
12/26/1936 - http://www.planecrashinfo.com/1936/1936-49.htm Undetermined aboard & fatalities.<br>
02/14/1942 - No info<br>
10/01/1942 - No info<br>
11/09/1944 - https://www.baaa-acro.com/crash/crash-consolidated-b-24-liberator-near-seljord-10-killed 10/10 killed. Also, info about the plane (B-24) and accident description. Not a war related loss but an accident.<br>
03/18/1946 - http://www.planecrashinfo.com/1946/1946-15.htm Undetermined aboard & fatalities.<br>
12/25/1946 - Several matches, no complete match with the information we have (none of them have 4 ground casualties). <br>
04/20/1957 - http://www.planecrashinfo.com/1957/1957-21.htm Undetermined aboard, one fatality.<br>
11/03/1977 - http://www.airsafe.com/events/airlines/elal.htm One fatality<br>
12/16/1981 - https://aviation-safety.net/wikibase/33132 12/12 fatalities. The year is incorrect according to at least 2 different sources - happened in 1980 at 16:15

With this recovered information we can fill out the missing info a decide what to do with the rows for which we have no information. <br>We need to take into consideration the historic context - the missing information is all from one source, the China National Aviation Corporation. This airliner was part of the Nationalist Chinese goverment before they lost the war, so the information (specially in the last years) is expected to be unclear. <br>Since we are talking about 7 rows, we can also drop them so we have no other missing values for fatalities in the dataset, and then fill out the new information obtained.

In [31]:
airplanes = airplanes.drop([airplanes.index[333], airplanes.index[364], airplanes.index[423], airplanes.index[569], airplanes.index[588], airplanes.index[763], airplanes.index[827]])
airplanes = airplanes.reset_index(drop=True)

In [32]:
airplanes_nopassengers = airplanes[(airplanes['aboard'].isna())]
airplanes_nopassengers

date time               location                        operator  \
26    10/20/1919  NaN        English Channel   Aircraft Transport and Travel   
347   03/07/1935  NaN  Schievelbein, Germany                        Deruluft   
668   11/09/1944  NaN        Seljord, Norway  Military - U.S. Army Air Corps   
1467  04/20/1957  NaN          Jirkouk, Iraq                      Air France   
2995  11/03/1977  NaN   Belgrade, Yugoslavia                           El Al   
3311  12/16/1981  NaN   Kuala Belait, Brunei             Bristow Helicopters   

     flight_no                            route                          type  \
26         NaN                              NaN             De Havilland DH-4   
347        NaN                              NaN              Rochrbach Roland   
668        NaN                              NaN                           NaN   
1467       NaN  Tehran, Iran - Istanbul, Turkey  Lockheed Super Constellation   
2995       NaN                              NaN                  Boeing B-747   
3311       NaN                              NaN             Aerospatiale Puma   

     registration cn/in  aboard  fatalities  ground  \
26         G-EAHG   NaN     NaN         NaN     NaN   
347        D-AJYP    45     NaN         3.0     0.0   
668      42-52196   NaN     NaN         NaN     NaN   
1467       F-BGNE  4514     NaN         1.0     0.0   
2995          NaN   NaN     NaN         1.0     0.0   
3311       9M-SSC  1481     NaN        12.0     0.0   

                                                summary  
26                                                  NaN  
347                                   Fuselage failure.  
668                                                 NaN  
1467  Explosive decompression. A passenger was sucke...  
2995                           Explosive decompression.  
3311                                                NaN

In [33]:
airplanes['aboard'][26] = 1
airplanes['fatalities'][26] = 0
airplanes['ground'][26] = 0
airplanes['summary'][26] = '''Crashed into the sea while attempting to land in fog.'''

airplanes['aboard'][347] = 13
airplanes['fatalities'][347] = 13
airplanes['summary'][347] = '''The DC-2, named "Gaai" operated on a passenger service from Milan, Italy to Amsterdam, the Netherlands. The aircraft took off at 11:36 hours, 
bound for Frankfurt, Germany, which was the next planned stop. Cruising at 5000 m altitude, ice accretion forced the crew to descend. At 3000 m the flight was out of 
icing conditions. However the aircraft was now flying between clouds shrouded mountains. Attempting to navigate visually, the flight continued at low altitude. 
Likely the crew entered the wrong mountain pass. They circled a valley, looking for a way out but low clouds and rain made it very difficult to continue flight.
The captain likely decided to perform a gear-up forced landing in the valley. Flaps were selected down and engine power was decreased. In a left hand turn the aircraft 
stalled and impacted the ground. The four crew members and the nine passengers died.'''

airplanes['aboard'][668] = 10
airplanes['fatalities'][668] = 10
airplanes['ground'][668] = 0
airplanes['type'][668] = '''Consolidated B-24 Liberator'''
airplanes['summary'][668] = ''' En route, the crew encountered poor weather and icing conditions. While all engines and both wings were contaminated by ice, 
the aircraft was unable to maintain the prescribed altitude and hit the slope of Mt Skorve located in the region of Seljord. All ten crew members were killed. '''

airplanes['aboard'][3311] = 12
airplanes['fatalities'][3311] = 12
airplanes['summary'][3311] = '''A second stage epicyclic module planetary gear fatigue failure caused loss of the main rotor, which severed the tail, causing the aircraft 
to crash in a swamp near the border settlement of Kuala Belait. The gearbox had a recent history of metallic debris being found on the magnetic chip detector
in the main module. The aircraft was on contract to Sarawak Shell. '''

/tmp/ipykernel_45867/3906361751.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['aboard'][26] = 1
/tmp/ipykernel_45867/3906361751.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['fatalities'][26] = 0
/tmp/ipykernel_45867/3906361751.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['ground'][26] = 0
/tmp/ipykernel_45867/3906361751.py:4: SettingWithCopyWarning: 
A value is trying to be set on a cop

In [34]:
airplanes_nopassengers = airplanes[(airplanes['aboard'].isna())]
airplanes_nopassengers

date time              location    operator flight_no  \
1467  04/20/1957  NaN         Jirkouk, Iraq  Air France       NaN   
2995  11/03/1977  NaN  Belgrade, Yugoslavia       El Al       NaN   

                                route                          type  \
1467  Tehran, Iran - Istanbul, Turkey  Lockheed Super Constellation   
2995                              NaN                  Boeing B-747   

     registration cn/in  aboard  fatalities  ground  \
1467       F-BGNE  4514     NaN         1.0     0.0   
2995          NaN   NaN     NaN         1.0     0.0   

                                                summary  
1467  Explosive decompression. A passenger was sucke...  
2995                           Explosive decompression.

We only have 2 remaining Nas for the aboard column. Not so relevant as we are focused more on the number of casualties. Both are decompression incidents so we can also calculate the ratio of these compared to the total of accidents.

<h3>Fatalities<h3>

Without any doubt, the worst possible issue that an airline company can face, even if the deaths are related to an underlying medical condition of the passenger, the mere association to the flight industry is a grime reminder of the human fragility.

In [35]:
airplanes['fatalities'].isna().sum()

0

We don't have any missing values. Looks like the ones already filled in manually and the rows we dropped with no information has completed this column as well. We will explore now the ground column.

<h3>Ground<h3>

Ground, when talking about aircraft accidents, is how we refer to the casualties derived from the accident that are not inside the aircraft, like ground personnel or, in case with crashes with buildings, people inside those.

In [36]:
airplanes['ground'].isna().sum()

11

We have 11 missing values. Let's take a look and find out if we can fill those out too.

In [55]:
airplanes_ground = airplanes[(airplanes['ground'].isna())]
airplanes_ground

date time                           location  \
310   11/20/1933  NaN                 Near Tsinan, China   
5250  05/26/2009  NaN  Near Isiro, DemocratiRepubliCongo   

                                 operator flight_no             route  \
310   China National Aviation Corporation       NaN  Canton - Shanghi   
5250                          Service Air       NaN      Goma - Isiro   

               type registration cn/in  aboard  fatalities  ground  \
310          Sinson          NaN   NaN     8.0         8.0     NaN   
5250  Antonov An-26       9Q-CSA  5005     4.0         4.0     NaN   

                                                summary  
310   Crashed into the Chingshan mountain range in fog.  
5250  The cargo plane crashed while on approach to I...

We can find very relevant information on the summary column for these (only one has no summary).

In [38]:
airplanes_ground['summary'][228]

'Crashed into a mountainside at an altitude of 4,500  feet during a snowstorm.'

We can safely assume no ground casualties at that height.

In [40]:
airplanes_ground['summary'][308]

'Crashed in a thickly wooded area upon taking off after the pilot became lost in fog.'

Same as before.

In [41]:
airplanes_ground['summary'][310]

'Crashed into the Chingshan mountain range in fog.'

We will assume 0 ground casualties

In [42]:
airplanes_ground['summary'][523]

nan

Since it is the only one missing the summary, we will try to find more info online.

https://en.wikipedia.org/wiki/List_of_accidents_and_incidents_involving_airliners_in_the_United_Kingdom#1930%E2%80%931939

Airspeed Oxford G-AFFM being operated by British Airways crashed at Gosport, Hampshire after it hit a barrage balloon cable, two crew killed.

This matches the info we have about aboard and fatalities. The date, model and registration perfectly matches so we will update the info accordingly.

In [43]:
airplanes_ground['summary'][633]

'Disappeared while en route on a non-combat mission. Wreckage found 39 years later on 4/30/1983.'

We will assume no ground casualties.

In [44]:
airplanes_ground['summary'][1165]

'Hit power lines and crashed into houses. Seventeen killed including civilians.'

Bad luck, we will have to look up more info, although B29s usually had a crew of 10-15 members. If the accident details cannot be found, we will consider 2-7 ground casualties.

https://aviation-safety.net/wikibase/85403<br>
Info found, 13/13 crew members, 4 ground casualties. We will also update description to this one as it is more complete:<br><br>
Destroyed during combat operations 7 February 1952: Shortly after takeoff from Yokota AB, while climbing in snow falls, the heavy bomber went out of control<br> and crashed in a huge explosion on several houses located about three miles (5 km) northwest of the airfield. All 13 crew members and five people on the ground were killed. 

In [45]:
airplanes_ground['summary'][1174]

'While on a training mission and flying blind on instruments the planes collided. One plane struck the ground and disintegrated. The other glided down several miles away, exploded and burned. Both planes crashed on ranches several miles apart about 18 miles from San Antonio . Six killed on one plane and seven on the other.'

Even though they crashed into ranches, summary reports only the aircraft casualties. We will assume 0 ground fatalities.

In [46]:
airplanes_ground['summary'][1810]

'Twelve killed, including civilians. Two civilian houses burnt.'

We need to find out either how many civilians or how many crew members died. 

https://www.town.kadena.okinawa.jp/kadena/P08_base%20digest_English.pdf<br>
"A KB50 aerial tanker crashed on take-off into Kadena’s Yara district. 2 were killed, 8 injured, and three homes completely destroyed in the ensuing fire."<br>
Looks like the website refers only to civilian casualties, so we will correct them accordingly.

In [47]:
airplanes_ground['summary'][4691]

'Crashed while attempting to land in poor weather.'

We can assume 0 ground casualties.

In [48]:
airplanes_ground['summary'][4695]

"Shortly after takeoff the aircraft impacted another plane and crashed to the ground. The failure of the pilot's of both airplanes to maintain a visual lookout while climbing and maneuvering resulting in an in-flight collision and subsequent collision with residences and terrain."

We need to dig in deeper.

http://edition.cnn.com/2000/US/06/23/plane.crash.03/index.html<br>
4 dead, 3 from this aircraft and another from the other aircraft they collided with. No ground casualties.

In [49]:
airplanes_ground['summary'][5250]

'The cargo plane crashed while on approach to Isiro-Matari Airport.'

We can assume no ground fatalities, just the crew. This concludes the exploration so we will now correct the values.

In [50]:
airplanes['ground'][228] = 0

airplanes['ground'][308] = 0

airplanes['ground'][310] = 0

airplanes['ground'][523] = 0
airplanes['summary'][523] = '''Airspeed Oxford G-AFFM being operated by British Airways crashed at Gosport, Hampshire after it hit a barrage balloon cable, two crew killed.'''

airplanes['ground'][633] = 0

airplanes['aboard'][1165] = 13
airplanes['fatalities'][1165] = 13
airplanes['ground'][1165] = 5
airplanes['summary'][1165] = '''Destroyed during combat operations 7 February 1952: Shortly after takeoff from Yokota AB, while climbing in snow falls, the heavy bomber went out of control 
and crashed in a huge explosion on several houses located about three miles (5 km) northwest of the airfield. All 13 crew members and five people on the ground were killed. '''

airplanes['ground'][1174] = 0

airplanes['aboard'][1810] = 10
airplanes['fatalities'][1810] = 10
airplanes['ground'][1810] = 2

airplanes['ground'][4691] = 0

airplanes['ground'][4695] = 0

airplanes['ground'][5250] = 0

/tmp/ipykernel_45867/3400115124.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['ground'][228] = 0
/tmp/ipykernel_45867/3400115124.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['ground'][308] = 0
/tmp/ipykernel_45867/3400115124.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airplanes['ground'][523] = 0
/tmp/ipykernel_45867/3400115124.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy

In [51]:
airplanes.reset_index(drop=True)

date   time                                           location  \
0     09/17/1908  17:18                                Fort Myer, Virginia   
1     07/12/1912  06:30                            AtlantiCity, New Jersey   
2     08/06/1913    NaN                 Victoria, British Columbia, Canada   
3     09/09/1913  18:30                                 Over the North Sea   
4     10/17/1913  10:30                         Near Johannisthal, Germany   
...          ...    ...                                                ...   
5249  05/20/2009  06:30                             Near Madiun, Indonesia   
5250  05/26/2009    NaN                  Near Isiro, DemocratiRepubliCongo   
5251  06/01/2009  00:15  AtlantiOcean, 570 miles northeast of Natal, Br...   
5252  06/07/2009  08:30       Near Port Hope Simpson, Newfoundland, Canada   
5253  06/08/2009    NaN                  State of Arunachal Pradesh, India   

                             operator flight_no  \
0                Military - U.S. Army       NaN   
1                Military - U.S. Navy       NaN   
2                             Private         -   
3              Military - German Navy       NaN   
4              Military - German Navy       NaN   
...                               ...       ...   
5249  Military - Indonesian Air Force       NaN   
5250                      Service Air       NaN   
5251                       Air France       447   
5252                       Strait Air       NaN   
5253      Military - Indian Air Force       NaN   

                                          route  \
0                                 Demonstration   
1                                   Test flight   
2                                           NaN   
3                                           NaN   
4                                           NaN   
...                                         ...   
5249                           Jakarta - Maduin   
5250                               Goma - Isiro   
5251                     Rio de Janeiro - Paris   
5252  Lourdes de BlanSablon - Port Hope Simpson   
5253                         Mechuka for Jorhat   

                                  type registration cn/in  aboard  fatalities  \
0                     Wright Flyer III          NaN     1     2.0         1.0   
1                            Dirigible          NaN   NaN     5.0         5.0   
2                     Curtiss seaplane          NaN   NaN     1.0         1.0   
3               Zeppelin L-1 (airship)          NaN   NaN    20.0        14.0   
4               Zeppelin L-2 (airship)          NaN   NaN    30.0        30.0   
...                                ...          ...   ...     ...         ...   
5249           Lockheed C-130 Hercules       A-1325  1982   112.0        98.0   
5250                     Antonov An-26       9Q-CSA  5005     4.0         4.0   
5251                   Airbus A330-203       F-GZCP   660   228.0       228.0   
5252  Britten-Norman BN-2A-27 Islander      C-FJJR    424     1.0         1.0   
5253                     Antonov An-32          NaN   NaN    13.0        13.0   

      ground                                            summary  
0        0.0  During a demonstration flight, a U.S. Army fly...  
1        0.0  First U.S. dirigible Akron exploded just offsh...  
2        0.0  The first fatal airplane accident in Canada oc...  
3        0.0  The airship flew into a thunderstorm and encou...  
4        0.0  Hydrogen gas which was being vented was sucked...  
...      ...                                                ...  
5249     2.0  While on approach, the military transport cras...  
5250     NaN  The cargo plane crashed while on approach to I...  
5251     0.0  The Airbus went missing over the AtlantiOcean ...  
5252     0.0  The air ambulance crashed into hills while att...  
5253     0.0  The military transport went missing while en r...  

[5254 rows x 13 columns]

In [52]:
airplanes.head()

date   time                            location  \
0  09/17/1908  17:18                 Fort Myer, Virginia   
1  07/12/1912  06:30             AtlantiCity, New Jersey   
2  08/06/1913    NaN  Victoria, British Columbia, Canada   
3  09/09/1913  18:30                  Over the North Sea   
4  10/17/1913  10:30          Near Johannisthal, Germany   

                 operator flight_no          route                    type  \
0    Military - U.S. Army       NaN  Demonstration        Wright Flyer III   
1    Military - U.S. Navy       NaN    Test flight               Dirigible   
2                 Private         -            NaN        Curtiss seaplane   
3  Military - German Navy       NaN            NaN  Zeppelin L-1 (airship)   
4  Military - German Navy       NaN            NaN  Zeppelin L-2 (airship)   

  registration cn/in  aboard  fatalities  ground  \
0          NaN     1     2.0         1.0     0.0   
1          NaN   NaN     5.0         5.0     0.0   
2          NaN   NaN     1.0         1.0     0.0   
3          NaN   NaN    20.0        14.0     0.0   
4          NaN   NaN    30.0        30.0     0.0   

                                             summary  
0  During a demonstration flight, a U.S. Army fly...  
1  First U.S. dirigible Akron exploded just offsh...  
2  The first fatal airplane accident in Canada oc...  
3  The airship flew into a thunderstorm and encou...  
4  Hydrogen gas which was being vented was sucked...

In [54]:
airplanes.isna().sum()

date               0
time            2205
location          19
operator          18
flight_no       4187
route           1695
type              25
registration     333
cn/in           1220
aboard             2
fatalities         0
ground             2
summary          383
dtype: int64

Time column is not so relevant in this case as we have many missing values. It is simply not feasible to fill out those values by hand and we can't drop so many rows either (almost 50% of the dataset) so we will drop the column.

In [56]:
airplanes = airplanes.drop(['time'], axis=1)
airplanes.head()

date                            location                operator  \
0  09/17/1908                 Fort Myer, Virginia    Military - U.S. Army   
1  07/12/1912             AtlantiCity, New Jersey    Military - U.S. Navy   
2  08/06/1913  Victoria, British Columbia, Canada                 Private   
3  09/09/1913                  Over the North Sea  Military - German Navy   
4  10/17/1913          Near Johannisthal, Germany  Military - German Navy   

  flight_no          route                    type registration cn/in  aboard  \
0       NaN  Demonstration        Wright Flyer III          NaN     1     2.0   
1       NaN    Test flight               Dirigible          NaN   NaN     5.0   
2         -            NaN        Curtiss seaplane          NaN   NaN     1.0   
3       NaN            NaN  Zeppelin L-1 (airship)          NaN   NaN    20.0   
4       NaN            NaN  Zeppelin L-2 (airship)          NaN   NaN    30.0   

   fatalities  ground                                            summary  
0         1.0     0.0  During a demonstration flight, a U.S. Army fly...  
1         5.0     0.0  First U.S. dirigible Akron exploded just offsh...  
2         1.0     0.0  The first fatal airplane accident in Canada oc...  
3        14.0     0.0  The airship flew into a thunderstorm and encou...  
4        30.0     0.0  Hydrogen gas which was being vented was sucked...

We're done with the basic cleaning, now we will focus on gathering info from the description to determine possible accident causes.